feat: upgrade knex

mertmit opened the pull request:
## Change Summary

Re #3582 
- Upgrade knex to v2.2.0
- Install tedious as mssql client for knex
- Upgrade sqlite3 to v5.1.12
  - Dockerfile
    - upgraded builder to node:16.17.0-alpine3.15
    - upgraded runner to alpine:3.15
    - install node-gyp dependencies (required with upgrades)
- SQLite drop column -> use new alter column instead of alter table
- MSSQL fixed custom schema handling

## Change type

- [x] feat: (new feature for the user, not a new feature for build script)
- [x] chore: (updating grunt tasks etc; no production code change)

wingkwong commented on packages/nocodb/src/lib/db/sql-client/lib/mssql/MssqlClient.ts:
```diff
@@ -356,9 +356,7 @@ class MssqlClient extends KnexClient {
-      const exists = await this.sqlClient.schema.hasTable(
-        this.getTnPath(args.tn)
-      );
+      const exists = await this.sqlClient.schema.hasTable(args.tn);
```
what's the reason to remove `getTnPath` here?

wingkwong commented on packages/nocodb/src/lib/db/sql-client/lib/mssql/MssqlClient.ts:
```diff
@@ -396,9 +394,7 @@ class MssqlClient extends KnexClient {
-      result.data.value = await this.sqlClient.schema.hasTable(
-        this.getTnPath(args.tn)
-      );
+      result.data.value = await this.sqlClient.schema.hasTable(args.tn);
```
what's the reason to remove getTnPath here?

mertmit replied:
Thank you for pointing out, this looks like a wrong change

o1lab commented on packages/nocodb/src/lib/db/sql-client/lib/sqlite/SqliteClient.ts:
```diff
@@ -1527,14 +1528,12 @@ class SqliteClient extends KnexClient {
+            args.columns[i],
+            oldColumn,
+            downQuery
+          );
```
SQLite down query should not be possible.

o1lab commented on packages/nocodb/src/lib/db/sql-client/lib/sqlite/SqliteClient.ts:
```diff
@@ -1956,19 +1945,31 @@ class SqliteClient extends KnexClient {
+
+      let q3 = this.genQuery(`UPDATE ?? SET ?? = ??;`, [t, n.cn, `${o.cno}_old`], shouldSanitize);
+
+      let q4 = this.genQuery(`ALTER TABLE ?? DROP COLUMN ??;`, [t, `${o.cno}_old`], shouldSanitize);
```
What does this `_old` hold ? was this part of our meta earlier. If not, we need to see its documented and uniform across other clients. Can discuss offline.

mertmit replied:
Because of the limitations with alter on SQLite on first query we rename original column to `colName`_old then we recreate column and update all data using this backup column on next query (q4) we drop it.

Also I realized there should be a pragma foreign keys on | off surrounding the queries to support foreign keys, I will handle this situation as well.

o1lab commented:
@mertmit : pls lets get this in tomorrow after you resolve the comments.

pranavxc commented on packages/nocodb/src/lib/db/sql-client/lib/sqlite/SqliteClient.ts:
```diff
@@ -1544,22 +1543,18 @@ class SqliteClient extends KnexClient {
-      }
 
-      await Promise.all(
+      if (synchronous) {
```
Since `await Promise.all(` is removed the catch part won't handle the exceptions that happened within the map callback function.

On exception it may print `Uncaught Error`.


You can try the following to reproduce the  bug.


```js

(async () => {

  // without promise.await
  try {
    [1,2,3].map(async () =>{
      throw new Error('test')
    })
  }catch(e){
    console.log(e)
  }

  
  // with promise.await
  try {
    await Promise.all([1,2,3].map(async () =>{
      throw new Error('test')
    }))
  }catch(e){
    console.log(e)
  }
```

mertmit replied:
I changed this to a for loop which seems to act same with Promise.all

pranavxc replied:
It will run sequentially with for loop

pranavxc replied:
looks good now :+1:

mertmit commented:
@wingkwong I dropped getTnPath for all schema specific queries because schema.hasTable returns wrong result for [schema].[table], other functions seem to work (both syntax) but I dropped them as well to keep a convention so all of the raw queries use getTnPath, none of the schema queries use it.

wingkwong commented:
@mertmit how does knex know which schema should be used? (let's say users define a new one)

mertmit commented:
Passed all the tests, builds are working.
Ready for review: @pranavxc @wingkwong (bumping as this left behind)

pranavxc commented on packages/nocodb/src/lib/db/sql-client/lib/mssql/MssqlClient.ts:
```diff
@@ -1599,9 +1595,9 @@ class MssqlClient extends KnexClient {
     try {
-      const query = `CREATE TRIGGER ${args.trigger_name} on ${this.getTnPath(
+      const query = `CREATE TRIGGER [${args.trigger_name}] on [${this.getTnPath(
         args.tn
-      )} \n${args.timing} ${args.event}\n as\n${args.statement}`;
+      )}] \n${args.timing} ${args.event}\n as\n${args.statement}`;
```
You can replace it with `genQuery`

```js
this.genQuery(`CREATE TRIGGER ?? on ?? \n${args.timing} ${args.event}\n as\n${args.statement}`, [args.trigger_name, this.getTnPath(
        args.tn
      )])
```

pranavxc commented on packages/nocodb/src/lib/db/sql-client/lib/mssql/MssqlClient.ts:
```diff
@@ -1853,11 +1849,11 @@ class MssqlClient extends KnexClient {
           this.querySeparator() +
-          `CREATE TRIGGER ${this.schema}.${triggerName} ON ${this.schema}.${args.table_name} AFTER UPDATE
+          `CREATE TRIGGER [${this.schema}].[${triggerName}] ON [${this.schema}].[${args.table_name}] AFTER UPDATE
                   AS
                   BEGIN
                       SET NOCOUNT ON;
-                      UPDATE ${this.schema}.${args.table_name} Set ${column.column_name} = GetDate() where ${pk.column_name} in (SELECT ${pk.column_name} FROM Inserted)
+                      UPDATE [${this.schema}].[${args.table_name}] Set [${column.column_name}] = GetDate() where [${pk.column_name}] in (SELECT [${pk.column_name}] FROM Inserted)
```
Use genQuery

pranavxc commented on packages/nocodb/src/lib/db/sql-client/lib/pg/PgClient.ts:
```diff
@@ -2102,7 +2102,7 @@ class PGClient extends KnexClient {
       }
     }
-    await this.sqlClient.raw(upQuery);
+    if (upQuery != '') await this.sqlClient.raw(upQuery);
```
Replace `!=` with `!==`

pranavxc commented:
@mertmit please resolve conflicts

pranavxc commented on packages/nocodb/package.json:
```diff
@@ -121,11 +121,13 @@
     "slash": "^3.0.0",
     "socket.io": "^4.4.1",
-    "sqlite3": "5.0.0",
+    "sqlite3": "5.1.2",
```
Issue related to sqlite upgrade https://github.com/nocodb/nocodb/issues/1970 , will test the docker build later in linux and confirm thee same :+1:

pranavxc replied:
#### Verified docker in:

- [ ] Ubuntu
- [ ] MacOS
- [ ] Windows 

#### Verified executable in:

- [ ] Ubuntu
- [ ] MacOS
- [ ] Windows

pranavxc replied:
Executable failed in windows with the following error 

![image](https://user-images.githubusercontent.com/61551451/199915252-99601c0a-8184-4c16-a492-fc4581332e10.png)

pranavxc replied:
Probably we need to add binaries for the platform here - https://github.com/nocodb/nocodb-timely/tree/main/binaries/binding

pranavxc replied:
Executable failing in mac as well

<img width="1136" alt="Screenshot 2022-11-04 at 12 58 30 PM" src="https://user-images.githubusercontent.com/61551451/199917051-600f31aa-afcc-4c78-94e6-c0710660ad8d.png">

mertmit replied:
Should be resolved with merge of @pranavxc 's PR

pranavxc replied:
yes 👍🏽

bcakmakoglu commented on packages/nocodb/src/lib/db/sql-client/lib/mssql/MssqlClient.ts:
```diff
@@ -2144,6 +2143,132 @@ class MssqlClient extends KnexClient {
     return result;
   }
 
+  /**
+   *
+   * @param {Object} - args
+   * @param {String} - args.parentTable
+   * @param {String} - args.parentColumn
+   * @param {String} - args.childColumn
+   * @param {String} - args.childTable
+   * @returns {Promise<{upStatement, downStatement}>}
+   */
+   async relationCreate(args) {
```
Let me chime in real quick.

These JSDocs are incorrect as shown by Webstorm:
<img width="545" alt="image" src="https://user-images.githubusercontent.com/78412429/200110413-b433acdf-3e22-4442-a545-d96f7c5330ee.png">

`Parameter args is not described in JSDoc `

And it's kind of unnecessary to use JSDocs for typing when we're already using TypeScript that handles that.
It even limits the typing because you describe return values as `Object` when in reality they aren't just `Object` but something else, for example `Result`.

Imo: Remove the JSDocs comments.

mertmit replied:
Thank you for comment. I agree those feel like duplicates with ts types.
I derived JSDocs from KnexClient. As those are mostly implemented by Pranav, @pranavxc what do you think on this?

Note: even if we decide removing/updating them, it is better we create a separate chore PR and get this one in as this PR is already big enough

pranavxc replied:
We added them when the entire project was in Javascript, since we migrated to Typescript we can remove them. We can do the cleanup part in a separate PR.
